### PR TITLE
[ROCm] Fix for MI200 related performance issue with Resnet50 model.

### PR DESCRIPTION
**Pull request by rsanthanam-amd**

### CODE REVIEW

File: tensorflow/compiler/xla/service/gpu/ir_emitter.cc [hunk 1]
```diff
@@ -289,11 +291,51 @@ bool IrEmitter::MaybeEmitDirectAtomicOperation(
       }
     }
 
+    if (isAMDGPU) {
+      std::string arch = ir_emitter_context_->amdgpu_arch();
+      llvm::PointerType* output_address_type =
+          llvm::dyn_cast<llvm::PointerType>(output_address->getType());
```
Comment (cheshire) [inline tensorflow/compiler/xla/service/gpu/ir_emitter.cc:297]: `auto` preferred when type is duplicated

**Review by cheshire — CHANGES_REQUESTED** [review]

File: tensorflow/compiler/xla/service/gpu/ir_emitter.cc [hunk 2]
```diff
@@ -289,11 +291,51 @@ bool IrEmitter::MaybeEmitDirectAtomicOperation(
...
+      CHECK_NE(output_address_type, nullptr);
+
+      // todo: implement for F16 via global_atomic_fadd_pk2f16
```
[the remark this answers is not in the document]
Comment (rsanthanam-amd) [reply tensorflow/compiler/xla/service/gpu/ir_emitter.cc:300]: it does not generate atomic add.  it simply generates a CAS loop.  i will remove that comment.

File: tensorflow/compiler/xla/service/gpu/ir_emitter.cc [hunk 3]
```diff
@@ -289,11 +291,51 @@ bool IrEmitter::MaybeEmitDirectAtomicOperation(
...
+      bool atomic_add_supported = (element_type == F32);
+      if (atomic_add_supported) {
+        if (output_address_type->getPointerAddressSpace() != 3) {
+          // the compiler will only generate a global_atomic_fadd if the pointer
+          // is in global addrspace (1)
+          auto cast_output_ptr = b_.CreateAddrSpaceCast(
+             output_address,
+             llvm::PointerType::get(output_address_type->getPointerElementType(),
+                               /*AddressSpace=*/1));
+          AtomicRMW(llvm::AtomicRMWInst::FAdd, cast_output_ptr, source,
```
[the remark this answers is not in the document]
Comment (rsanthanam-amd) [reply tensorflow/compiler/xla/service/gpu/ir_emitter.cc:310]: in the if branch, we use the casted address space.  it the else branch we directly use the output address.

File: tensorflow/compiler/xla/service/gpu/ir_emitter.cc [hunk 4]
```diff
@@ -289,11 +291,51 @@ bool IrEmitter::MaybeEmitDirectAtomicOperation(
...
-      AtomicRMW(llvm::AtomicRMWInst::Add, output_address, source,
-                llvm::MaybeAlign(),
-                llvm::AtomicOrdering::SequentiallyConsistent);
+      if (isAMDGPU) {
+        AtomicRMW(llvm::AtomicRMWInst::Add, output_address, source,
+                  llvm::MaybeAlign(),
+                  llvm::AtomicOrdering::SequentiallyConsistent,
+                  b_.getContext().getOrInsertSyncScopeID("agent"));
+      } else {
+        AtomicRMW(llvm::AtomicRMWInst::Add, output_address, source,
```
[the remark this answers is not in the document]
Comment (rsanthanam-amd) [reply tensorflow/compiler/xla/service/gpu/ir_emitter.cc:335]: it is originally a default parameter but we need to change it to an AMDGPU specific sync scope.

File: tensorflow/compiler/xla/service/gpu/ir_emitter.cc (hunk 1 above)
Comment (rsanthanam-amd) [reply tensorflow/compiler/xla/service/gpu/ir_emitter.cc:297]: fixed

File: tensorflow/compiler/xla/service/gpu/ir_emitter.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (rsanthanam-amd) [reply tensorflow/compiler/xla/service/gpu/ir_emitter.cc]: merged into a single AtomicRMW

File: tensorflow/compiler/xla/service/gpu/ir_emitter.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (rsanthanam-amd) [reply tensorflow/compiler/xla/service/gpu/ir_emitter.cc]: fixed (less duplication).

File: tensorflow/compiler/xla/service/gpu/ir_emitter.cc [hunk 5]
```diff
@@ -471,10 +525,21 @@ Status IrEmitter::EmitAtomicOperationUsingCAS(const HloComputation& computation,
   // Emit code to perform the atomicCAS operation
   // (cas_old_output, success) = atomicCAS(memory_address, cas_old_output,
   //                                       cas_new_output);
-  llvm::Value* ret_value = AtomicCmpXchg(
-      atomic_memory_address, cas_old_output, cas_new_output, llvm::MaybeAlign(),
-      llvm::AtomicOrdering::SequentiallyConsistent,
-      llvm::AtomicOrdering::SequentiallyConsistent);
+  llvm::Value* ret_value = [&]() {
+    llvm::Triple target_triple = llvm::Triple(module_->getTargetTriple());
+    if (target_triple.isAMDGPU()) {
```
[the remark this answers is not in the document]
Comment (rsanthanam-amd) [reply tensorflow/compiler/xla/service/gpu/ir_emitter.cc]: done.

File: tensorflow/compiler/xla/service/gpu/ir_emitter.cc [hunk 6]
```diff
@@ -289,11 +291,51 @@ bool IrEmitter::MaybeEmitDirectAtomicOperation(
       }
     }
 
+    if (isAMDGPU) {
+      std::string arch = ir_emitter_context_->amdgpu_arch();
```
[the remark this answers is not in the document]
Comment (rsanthanam-amd) [reply tensorflow/compiler/xla/service/gpu/ir_emitter.cc:295]: it is cleaner now so i am hoping to avoid moving into a separate function.

**Comment by rsanthanam-amd** [comment]
@cheshire Please review the changes i have made based on your comments.

i need to add more changes in a separate commit because the AMD ROCm Community CI build is failing.

File: tensorflow/compiler/xla/service/gpu/ir_emitter.cc [hunk 7]
```diff
@@ -289,11 +291,42 @@ bool IrEmitter::MaybeEmitDirectAtomicOperation(
       }
     }
 
+    if (isAMDGPU) {
+      auto output_address_type =
```
Comment (cheshire) [inline tensorflow/compiler/xla/service/gpu/ir_emitter.cc:295]: Due to nesting, this gets hard to read, maybe factor out the AMD emission to a separate function? Also `if (IsEmittingForAMDGPU())` seems more readable.

File: tensorflow/compiler/xla/service/gpu/ir_emitter.cc [hunk 8]
```diff
@@ -315,7 +350,9 @@ bool IrEmitter::MaybeEmitDirectAtomicOperation(
                       ? llvm::AtomicRMWInst::Min
                       : llvm::AtomicRMWInst::UMin;
     AtomicRMW(opcode, output_address, source, llvm::MaybeAlign(),
-              llvm::AtomicOrdering::SequentiallyConsistent);
+              llvm::AtomicOrdering::SequentiallyConsistent,
+              (isAMDGPU) ? b_.getContext().getOrInsertSyncScopeID("agent") :
```
Comment (cheshire) [inline tensorflow/compiler/xla/service/gpu/ir_emitter.cc:354]: This ternary is used quite a few times, factor out into a function?

**Review by cheshire — CHANGES_REQUESTED** [review]

File: tensorflow/compiler/xla/service/gpu/ir_emitter.cc (hunk 7 above)
Comment (rsanthanam-amd) [reply tensorflow/compiler/xla/service/gpu/ir_emitter.cc]: done.

File: tensorflow/compiler/xla/service/gpu/ir_emitter.cc (hunk 8 above)
Comment (rsanthanam-amd) [reply tensorflow/compiler/xla/service/gpu/ir_emitter.cc]: done.

File: tensorflow/compiler/xla/service/gpu/ir_emitter.cc [hunk 9]
```diff
@@ -509,6 +515,50 @@ Status IrEmitter::EmitAtomicOperationForNestedComputation(
                                      source_address);
 }
 
+bool IrEmitter::IsEmittingForAMDGPU() const {
+  llvm::Triple target_triple = llvm::Triple(module_->getTargetTriple());
+  return target_triple.isAMDGPU();
+}
+
+bool IrEmitter::EmitAMDGPUAtomics(
```
Comment (rsanthanam-amd) [inline tensorflow/compiler/xla/service/gpu/ir_emitter.cc:523]: Please let me know if you have a better name for this function.

**Comment by rsanthanam-amd** [comment]
@cheshire gentle ping

File: tensorflow/compiler/xla/service/gpu/ir_emitter.cc [hunk 10]
```diff
@@ -509,6 +515,50 @@ Status IrEmitter::EmitAtomicOperationForNestedComputation(
...
+    llvm::Value* output_address, llvm::Value* source,
+    PrimitiveType element_type) {
+  if (IsEmittingForAMDGPU()) {
```
Comment (cheshire) [inline tensorflow/compiler/xla/service/gpu/ir_emitter.cc:526]: I think this is a bit misleading vs. the function name (it says "emits", but in practice only emits for AMD GPUs). Could we make the fact that we are emitting for AMD GPU a precondition with a CHECK and move the test (IsEmitting..()) to the caller?

File: tensorflow/compiler/xla/service/gpu/ir_emitter.cc [hunk 11]
```diff
@@ -509,6 +515,50 @@ Status IrEmitter::EmitAtomicOperationForNestedComputation(
...
+          return b_.CreateAddrSpaceCast(
+              output_address,
+              llvm::PointerType::get(output_address_type->getPointerElementType(),
+                                     /*AddressSpace=*/1));
+        } else {
+          // adds to shared memory are always atomic.
+          return output_address;
+        }
+      }();
+      AtomicRMW(llvm::AtomicRMWInst::FAdd, output_ptr, source,
```
Comment (cheshire) [inline tensorflow/compiler/xla/service/gpu/ir_emitter.cc:546]: What is the main difference between NVidia and AMD for atomic calls? Is it just the casting? Maybe we could share at a lower granularity?

**Review by cheshire — CHANGES_REQUESTED** [review]
Sorry a few more nits!

File: tensorflow/compiler/xla/service/gpu/ir_emitter.cc (hunk 11 above)
Comment (rsanthanam-amd) [reply tensorflow/compiler/xla/service/gpu/ir_emitter.cc:546]: it's not just the casting.  it is also the sync scope.  i would imagine that nvidia would simply use the default 'system' sync scope but we need to use 'agent' for performance reasons.

Comment (rsanthanam-amd) [reply tensorflow/compiler/xla/service/gpu/ir_emitter.cc:546]: i think the 'agent' sync scope is AMD specific.

File: tensorflow/compiler/xla/service/gpu/ir_emitter.cc (hunk 10 above)
Comment (rsanthanam-amd) [reply tensorflow/compiler/xla/service/gpu/ir_emitter.cc]: done.  i think the intent has been clarified.

**Comment by rsanthanam-amd** [comment]
@cheshire gentle ping.

File: tensorflow/compiler/xla/service/gpu/ir_emitter.cc [hunk 12]
```diff
@@ -509,6 +519,40 @@ Status IrEmitter::EmitAtomicOperationForNestedComputation(
...
+          output_address,
+          llvm::PointerType::get(output_address_type->getPointerElementType(),
+                                 /*AddressSpace=*/1)) :
+      // adds to shared memory are always atomic.
+      output_address;
+
+  AtomicRMW(llvm::AtomicRMWInst::FAdd, output_ptr, source,
+            llvm::MaybeAlign(),
+            llvm::AtomicOrdering::SequentiallyConsistent,
+            b_.getContext().getOrInsertSyncScopeID("agent"));
```
Comment (cheshire) [inline tensorflow/compiler/xla/service/gpu/ir_emitter.cc:547]: ..but in other place you are using same AtomicRMW call for both and DetermineSyncScope as a last argument? Would it make sense to do the same here?

Comment (rsanthanam-amd) [reply tensorflow/compiler/xla/service/gpu/ir_emitter.cc:547]: well, that piece of code is already guarded by IsEmittingforAMDGPU so hard coding the sync scope to agent is just right.

For the other places, i think we need to use default for CUDA and agent for AMDGPU.

Comment (cheshire) [reply tensorflow/compiler/xla/service/gpu/ir_emitter.cc:547]: Right, sorry for nitpicking, I've meant whether we need a whole separate function to emit atomics?

Comment (rsanthanam-amd) [reply tensorflow/compiler/xla/service/gpu/ir_emitter.cc:547]: @cheshire Now I understand your comment.  There was a typo in my PR which I have now fixed.

**Review by cheshire — APPROVED** [review]

**Pull request merged**